### ISSUE TITLE
60 min Token Exipiration

### ISSUE THREAD

**Issue by bmrkobrad** [pluginsGLPI/oauthimap#8]
The script works, but when create authorization, in our case for Gmail, it expires after 60 min. After expiration it is necessary to delete the existing authorization, create a new one and set the Receiver on Active -> Yes.
It is very difficult to keep helpdesk functional in this way.
Is there any way to increase the token time span, or refresh the token automatically?

**Comment by bpain** [comment]
> 
> 
> The script works, but when create authorization, in our case for Gmail, it expires after 60 min. After expiration it is necessary to delete the existing authorization, create a new one and set the Receiver on Active -> Yes.
> It is very difficult to keep helpdesk functional in this way.
> Is there any way to increase the token time span, or refresh the token automatically?

Can confirm!

**Comment by alasfarg** [comment]
Yes, it is the same here and the collector appears in error quite soon with this message 
Adresse de courriel invalide
Une erreur est survenue en essayer de se connecter au collecteur.
Required parameter not passed: "refresh_token"

I have not been able to find where is the place of this parameter

**Comment by Weastbug** [comment]
Same for me,

Problem describe in GLPI forum too https://forum.glpi-project.org/viewtopic.php?id=280101

Here is my explain :

Same problem here for google too.

Oauth2 Autorisation Expiration is for 1 hour when created.

Collector work fine in this hour, after this i have a refresh token error. Need delete and recreate the authorisation again and reactive collector for have it working back.
```
Adresse de courriel invalide
Une erreur est survenue en essayer de se connecter au collecteur.
Required parameter not passed: "refresh_token"
```

~~Based on my search and if i have understand GLPI Oauth need do the query with access_type=offline :~~

~~> [https://stackoverflow.com/questions/53379222/google-api-how-to-increase-access-token-expiration-date]()~~
~~> Due to security reasons, you cannot change the duration of the access token's expiry. However, you can refresh an access token without prompting the user for permission if you requested offline access to the scopes associated with the token.~~
> 
~~> - If you use a Google API Client Library, the client object refreshes the access token as needed as long as you configure that object for offline access.~~
~~> - If you are not using a client library, you need to set the access_type HTTP query parameter to offline when redirecting the user to Google's OAuth 2.0 server. In that case, Google's authorization server returns a refresh token when you exchange an authorization code for an access token. Then, if the access token expires (or at any other time), you can use a refresh token to obtain a new access token.~~

Edit: the accessType in query was already set as offline..., don't understand all...
```
GLPI\marketplace\oauthimap\inc\application.class.php
         case Google::class:
            $params['accessType'] = 'offline';
            break;
```
Here sample of url on auth creation : `https://accounts.google.com/o/oauth2/v2/auth/oauthchooseaccount?scope=openid%20email%20profile%20https%3A%2F%2Fmail.google.com%2F&access_type=offline&state=xxxxxxxxxxx&response_type=code&redirect_uri=http%3A%2F%2Fxxxxxxxxxxx%2Fmarketplace%2Foauthimap%2Ffront%2Fauthorization.callback.php&client_id=xxxxxxxxxxx.apps.googleusercontent.com&flowName=GeneralOAuthFlow`

Thanks in advance.

**Comment by Weastbug** [comment]
> After expiration it is necessary to delete the existing authorization, create a new one and set the Receiver on Active -> Yes

Don't need to delete. Just create it again and then don't need to set Active to Yes ;)

**Comment by Info-NI** [comment]
Hello,

We have the same probleme.

Thanks in advance

**Comment by bmrkobrad** [comment]
If update collector without deleting, the effect is the same with one step less. The token expires after one hour.
But there are also good news. If use Gmail, you don't need this plugin at all. You can still use the old way for IMAP authentication.
It is necessary to enable 2-step verification for the account, and you can generate an app specific password. It works with IMAP receiver in GLPI.
I don't know how it is with Azure.
